### ISSUE TITLE
Inconsistent environment and behaviour

### ISSUE THREAD

**Issue by lloeki** [atom/atom#4126]
# Problem

Various tools don't operate correctly depending on the order and the manner with which Atom is started.

Also, persisted state is not restored consistently.
# Cause

Depending the manner in which Atom is started, a different environment is inherited.
- **launchd event**: inherits launchd mostly blank env, which means no shell env vars
- **`atom` in the terminal**: inherits currently set vars

This applied to when the app is _started_, not when a new window is opened, in which case that window will inherit the current process env (which is normal OS X behaviour)

I venture the creation of the `atom` command this way is on purpose to work around the lack of env when run from launchd. But please, dear reader, continue on to the next step.
# Current workaround / Anti-feature

Manually control Atom startup each time you want to use it in a specific environment.

Why is it an anti feature? If you want two projects open at the same time but with different environments, you're busted.

Also, long running sessions and _"wait what env was that started with already?"_ (session can also mean the idiomatic "document != app" OS X way: "Atom.app is started but no windows are opened") .

Also also, frustration quickly sets in when you start it up the wrong way and have to quit and restart halfway through your work and notice your linter does not actually run (manual config does not cut it because 1. it's manual and 2. constantly changing versions means constantly changing path).
# Possible solution
1. wrap Atom.app startup in a login shell (i.e make `Atom.app/Contents/MacOS/Atom` a script that calls whatever the user shell is† as a login, non-interactive one before calling the real `Atom` binary), to inherit global env, so that chruby/rbenv/GOPATH/whatever works.
2. make `atom` command send a signal to `Atom.app` instead of starting atom itself via `Contents/MacOS/Atom`. Basically, `atom foo/bar` has no reason to be different from `open -a Atom foo/bar`. ~~Precisely, it _does not_ inherit the local environment.~~
3. local environment management is bestowed to package implementor (meaning detection of `.git`, `Gemfile`, `.rubocop.yml`, `.gopath`, which they have to do anyway else at least starting via launchd is broken, if not more), not to manual handling by developer.
# Consequences
- `PATH` & al. is properly set and tools like linters can leverage the env to use chruby/rbenv/whatever.
- Clicking the icon, starting via spotlight, opening a file from Finder, using `open -a Atom` and using `atom foo` on the terminal are all consistent, whatever the order they are executed in.
- Everything is now consistent and we can now rejoice as glittery roses spontaneously bloom out of Bose-Einstein condensates.
# Additional note

Pet peeve: `atom` is currently equivalent to `atom .`. While I can see how this could be ok on other platforms, this is subtly wrong on OS X as `atom` is expected to be equivalent to `open -a Atom`, i.e start the app and don't open anything (unless it's set up to open a new document when no document is opened, otherwise merely focus the app, and if the app is not started, start it)

On another planet, I don't know if this issue applies to other OSes as well besides OS X. Everything depends on how process management and app/doc/task GUI model is handled there.
# Final word

All of this really matters daily when you juggle between projects and makes Atom feel like an unfinished, brittle product, making angry coworkers throwing nails at me for having pushed Atom on them.

† think fish-shell or (god forbid) csh.

**Comment by lloeki** [comment]
I overlooked this from `man open`:

     Opened applications inherit environment variables just as if you had
     launched the application directly through its full path.  This behavior
     was also present in Tiger.

**Comment by batjko** [comment]
Does this apply to Windows and Linux as well then?

**Comment by lloeki** [comment]
Testing with:

```javascript
new (require("atom").BufferedProcess)({
  command: "ruby",
  args: ["-e", 'puts ENV.map(&:to_s).join("\n")'],
  stdout: function(out) { console.log("OUT:\n", out); },
  stderr: function(out) { console.log("ERR:\n", out); },
});
```

Output on `atom` and `open -a Atom` don't differ (which is normal given `atom.sh` content) while output of either one with an icon launch differ significantly, the latter missing a lot of vars:

```diff
< ["RBENV_VERSION", "2.2.2"]
< ["TERM_PROGRAM", "Apple_Terminal"]
< ["ANDROID_HOME", "/usr/local/opt/android-sdk"]
< ["TERM", "xterm-256color"]
---
>  ["RBENV_VERSION", "2.2.2"]
6d2
< ["CLICOLOR", "1"]
9,10d4
< ["TERM_PROGRAM_VERSION", "360"]
< ["TERM_SESSION_ID", "C06FFFD2-16C3-48CD-AE1D-9CF8A31C6DA2"]
18c12
< ["PATH", "/Users/lloeki/.rbenv/versions/2.2.2/bin:/usr/local/Cellar/rbenv/0.4.0/libexec:/Users/lloeki/.rbenv/shims:/Users/lloeki/.rbenv/bin:/usr/local/bin:/usr/local/sbin:/Users/lloeki/.local/bin:/Users/lloeki/.rbenv/shims:/Users/lloeki/.rbenv/bin:/Users/lloeki/Library/Python/2.7/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/share/npm/bin:/usr/local/opt/go/libexec/bin"]
---
> ["PATH", "/Users/lloeki/.rbenv/versions/2.2.2/bin:/usr/local/Cellar/rbenv/0.4.0/libexec:/Users/lloeki/.rbenv/shims:/Users/lloeki/.rbenv/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin"]
20,21d13
< ["EDITOR", "vim"]
< ["LANG", "en_GB.UTF-8"]
26c18
< ["SHLVL", "2"]
---
> ["SHLVL", "0"]
28d19
< ["GREP_OPTIONS", "--color=auto"]
31,32d21
< ["VISUAL", "vim"]
< ["HOMEBREW_NO_EMOJI", "you bet"]
```

Then, using the following hack:

```bash
mv /Applications/Atom.app/Contents/MacOS/Atom{,.bin}
cat > /Applications/Atom.app/Contents/MacOS/Atom <<<'#!/bin/bash -l
"$(dirname "${BASH_SOURCE[0]}")"/Atom.bin "$@"'
chmod +x /Applications/Atom.app/Contents/MacOS/Atom
```

Diffing the result of a "atom" invocation and a Spotlight opening of Atom:

```diff
2d1
< ["TERM_PROGRAM", "Apple_Terminal"]
4d2
< ["TERM", "xterm-256color"]
9,10d6
< ["TERM_PROGRAM_VERSION", "360"]
< ["TERM_SESSION_ID", "C06FFFD2-16C3-48CD-AE1D-9CF8A31C6DA2"]
21d16
< ["LANG", "en_GB.UTF-8"]
26d20
< ["SHLVL", "2"]
27a22
> ["SHLVL", "1"]
```

Updating the hack:

```bash
cat > /Applications/Atom.app/Contents/MacOS/Atom <<<'#!/bin/bash -l
[[ -z $LANG ]] && LANG=$(osascript -e "user locale of (get system info)" | sed -e "s/@.*//" -e "s/\$/.UTF-8/")
"$(dirname "${BASH_SOURCE[0]}")"/Atom.bin "$@"'
```

Diff becomes:

```diff
2d1
< ["TERM_PROGRAM", "Apple_Terminal"]
4d2
< ["TERM", "xterm-256color"]
9,10d6
< ["TERM_PROGRAM_VERSION", "360"]
< ["TERM_SESSION_ID", "C06FFFD2-16C3-48CD-AE1D-9CF8A31C6DA2"]
26d21
< ["SHLVL", "2"]
27a23
> ["SHLVL", "1"]
```

*Note*: The `sed` thing is because the `osascript` output looks like this on my machine (prefpane set as UK locale with custom currency):

```
$ osascript -e 'user locale of (get system info)'                                  
en_GB@currency=EUR
```

And here is a more complete less-of-a-hack that keeps using the actual user shell:

```
#!/bin/bash

if [[ -z $LANG ]]; then
    LANG=$(osascript -e "user locale of (get system info)" | sed -e "s/@.*//" -e "s/\$/.UTF-8/")
    export LANG
fi

user_shell="$(dscl . -read "/Users/$USER" UserShell | sed -e 's/UserShell: //')"
case "$user_shell" in
    */zsh|*/bash)
        "$user_shell" -l -c "$(dirname "${BASH_SOURCE[0]}")"/Atom.bin "$@"
        ;;
    *)
        echo "Unsupported shell: $user_shell"
        exit 1
        ;;
esac
```

**Comment by lloeki** [comment]
> Does this apply to Windows and Linux as well then?

@batjko very possible, although I'm unable to test. It all depends within which environment the system starts the Atom process and whether or not subsequent opened Atoms are "shared" as is the case on OS X.

From my experience on Linux I'd venture to say an Atom process started by a desktop environment will inherit whatever env the DE (and hence the desktop manager such as GDM) has set (`Xsession` and stuff, possibly `.xinitrc` for the most adventurous). This means that if people follow proper practices† that make `bashrc` the interactive script and `bash_profile` the login script, only `bash_profile` will be parsed and the content of `bashrc` will be missing to the DE, since it'll only be handled on interactive sessions and so, only as soon as a terminal is opened and within that terminal. For Windows I really don't know (I don't even know if there is an equivalent of `bash{rc,_profile}`.

The above proof-of-concept code makes Atom always run on OS X within an environment that has evaluated the user's login scripts (whatever the user shell is, as long as it's zsh or bash for now)

† Ever ran a non-interactive command over ssh such as `ssh 1.2.3.4 ls -l` on a Debian machine and have it output `stdin: is not a tty`? That's because they mixed things up and put a `mesg` in a `.profile` skel script.

**Comment by lloeki** [comment]
Apparently on OS X a little bit of ObjC code can do the equivalent of the `osascript` above:

```objc
NSLocale *locale; locale = [NSLocale currentLocale];
NSLog(@"Current locale Identifier %@",[locale localeIdentifier]); 
```

Might be useful to conditionally set `LANG` from within Atom's own startup code when it's not set.

**Comment by lloeki** [comment]
Updated wrapper with a few more shells. Not sure how `ksh` works.

**Comment by lloeki** [comment]
Ah, I forgot that I had this in my `init.coffee`, which means the unhacked diff is bigger:

```javascript
process.env.PATH = [
  "#{process.env.HOME}/.rbenv/shims"
  "#{process.env.HOME}/.rbenv/bin"
  "/usr/local/bin"
  "/usr/local/sbin"
  process.env.PATH
].join(':')
```

**Comment by lloeki** [comment]
I settled on this as a workaround for `LANG`:

```coffee
new (require("atom").BufferedProcess)
  command: "osascript"
  args: ["-e", "user locale of (get system info)"]
  stdout: (out) ->
    m = /^([a-z]{2}_[A-Z]{2})/.exec(out)
    process.env.LANG = "#{m[1]}.UTF-8" if m?
```

**Comment by mnquintana** [comment]
This looks like a duplicate of https://github.com/atom/atom/issues/6956 – closing in favor of that issue because I think it gets to the root cause a little more directly. Let's continue the discussion there.

**Comment by lloeki** [comment]
@mnquintana I kindly beg to differ: this is absolutely not a duplicate of #6956, as it's more general than that. It is a matter of general consistency, not of merely setting up `PATH` correctly in the event of being launched by launchd all the while being unconcerned about other env vars.

First, as a very realistic `PATH` example, consider the following:

- close atom
- venv1: `atom .`
- do stuff
# later
- venv2: `atom .`
- sad puppy venv2 has venv1 PATH
# later
- not_a_venv: `atom .`
- sad puppy has venv1 PATH even though it's not even in a venv

Alternatively, blindly inheriting env vars is most probably *not* a good idea in the case of Atom:

```bash
export RAILS_ENV=my_special_env
export SOME_MAGIC_TOKEN_FOR_THIS_PROJECT=deadbeef
export PATH=/foo:$PATH
rake whatevs
atom .
# do stuff
# done? close window but don't quit atom yet
# days later... another project
atom .
# WAT? stuff doesn't work / uses wrong token
```

As a consequence, per project env vars should definitely not be inherited from the starting env command (in the current architecture at least, due to its extreme side-effect-ey nature) and `https://atom.io/packages/atomenv` sounds much more promising and robust.

Regarding `PATH`, the suggestions I see in the linked issue:

- filtering on as volatile a thing as a pid of all things to find children, live-patching them possibly too late, and running a wheelbarrow load of `bash` synchronously, best described as a glorious hack
- running `bash` as non-login, which will omit stuff, and probably the most relevant one that is `bash_profile` since it's certainly the one supposed to set `PATH` (since it's ru non-interactively, bashrc shouldn't even be touched, see [this comment](https://github.com/atom/atom/issues/4126#issuecomment-132122283))
- using a `launchctl config` or `launchctl setenv` affecting every other app that is not Atom and possibly systemwide is both manual and not a serious option.

Still, since one won't be able to hardcode every possible var needed by some (public or homegrown) tool that Atom might run via `make`, `rake` or some Atom package, if one really wants to get "global" (as opposed to per-project) env vars into Atom from the user's dotfiles, the best course of action is either one of:
1. wrap Atom in `bash -l` (or equivalent)
2. run `bash -l set` once, parse the output, and set env internally as early as possible before any fork, every subsequent child process inheriting from that, by design

The proof-of-concept above is really just that, a PoC based on case 1, since case 2 may require diving into Electron (I really don't know, maybe not).

The alternative to obtain every env var from a bash login instance is to have a clean, minimal env such as the one set by launchd, and let the various scenarios (rbenv, pyenv, gopath) be handled by tools that will set and handle the relevant env vars as needed, each tool being aware of multiple project root per environment. In this scenario, #6956 is relevant but far from handling the whole scope of things.

Whether or not one wants to obtain every env var from a bash login instance, one has to clean up env by whitelisting allowed vars and unsetting the remainder (or something to that effect) so that every Atom instance is consistently started with the same env, whichever the method of starting up was.

Handling projects written in various concurrent versions of ruby, go python and node simultaneously, I'm being hit by those issues daily and had ample time (8 months, which ironically makes this issue predate #6956 by 6 months) to analyse the root cause of this issue, which is that launchd and a terminal's bash don't have the same env at all and that Atom keeps using the first one it got on startup.

cc @maxbrunsfeld which seemed to have an opinion on this in #8382.

**Comment by lloeki** [comment]
In addition to the `LANG` stuff above (which doesn't get set even by `bash -l`) and instead of the wrapper PoC I set this up in my `init.coffee` as a workaround:

```ruby
# this won't be copied over from the shell
envBlacklist = [
  # terminal info
  'TERM'
  'TERM_PROGRAM'
  'TERM_PROGRAM_VERSION'
  'TERM_SESSION_ID'
  'COLUMNS'
  'ROWS'
  # atom sets this itself
  'NODE_PATH'
  'GOOGLE_API_KEY'
  'ATOM_HOME'
]

# only this is allowed to stay, whether it comes from launchd/shell or shell copy
envWhitelist = [
  'SHELL'
  'TMPDIR'
  'Apple_PubSub_Socket_Render'
  'ATOM_HOME'
  'USER'
  'RBENV_ROOT'
  'SSH_AUTH_SOCK'
  '__CF_USER_TEXT_ENCODING'
  'RBENV_HOOK_PATH'
  'GOOGLE_API_KEY'
  'PATH'
  'PWD'
  'NODE_PATH'
  'XPC_FLAGS'
  'NODE_ENV'
  'XPC_SERVICE_NAME'
  'SHLVL'
  'HOME'
  'RBENV_DIR'
  'LOGNAME'
  'DISPLAY'
]

new (require("atom").BufferedProcess)
  command: "bash"
  args: ["-l", "-c", "set"]
  stdout: (out) ->
    # this could use some promise
    lines = out.split('\n')
    for line in lines
      m = /^([A-Za-z_]+)=(.*)$/.exec(line)
      # clean env
      for k of process.env
        delete process.env[k] if k not in envWhitelist
      # copy env
      if m? and m[1] not in envBlacklist and m[1] in envWhitelist
        process.env[m[1]] = m[2]
```

In a perfect world, there should be a promise or a message for which to subscribe to announcing the env is ready for use. An alternative is to get rid of the whitelist stuff, but this won't prevent unwanted var leak into subsequent atom instances.

**Comment by lloeki** [comment]
@mnquintana, thanks for reopening!

**Comment by joefitzgerald** [comment]
I think there is a very simple solution to this problem:

1. Inspect the PATH for the darwin platform for a value that indicates it was not launched via the shell
2. Run `process.env.SHELL` and then `env`, using the results to patch process.env with the updated environment

https://atom.io/packages/environment does this today, providing a service with a patched environment. We could adapt this approach such that it is run on startup?

A related issue is that if you launch atom with `SOMEVAR=FOO atom .` and then _subsequently_ launch `SOMEVAR=BAR atom .` the second invocation will cause a new workspace to be opened as a child of the original process, resulting in `SOMEVAR` continuing to have a value of `FOO` (when it should be `BAR` in the second window and `FOO` in the first window).

**Comment by lloeki** [comment]
@joefitzgerald if you read my original post and my comment [above](https://github.com/atom/atom/issues/4126#issuecomment-132192261), you'll see I covered all possible cases.

> Run process.env.SHELL and then env, using the results to patch process.env with the updated environment

This is what I do in my workaround above, except process.env.SHELL may be entirely unset on OS X, therefore one would have to use the `dscl` thing I described in another comment above.

> https://atom.io/packages/environment

Doesn't make things *consistent*. See below.

> A related issue is that if...

Again, covered by everything I said and my implementation above. The only solution is to whitelist stuff 
filter the rest, and let feature packages handle things (like go-plus), so that it is always consistent. This is simply mandatory because in your example you will *never ever* get `SOMEVAR=BAR` reaching Atom, and things would anyway just get even weirder with multi-project windows.

**Comment by joefitzgerald** [comment]
@lloeki would you mind testing #11054 and letting me know if it addresses your issue?

**Comment by lloeki** [comment]
@joefitzgerald I'll take a peek

**Comment by lloeki** [comment]
@joefitzgerald The following is as tested on current beta.

I understand #11054 fixes a use case by effectively:
- making each window a sandbox WRT `process.env`
- allow `atom` started instances to receive the environment (thanks to the above)

So, this use case works:
1. open Atom via dock/finder/spotlight
2. open a project/edit and save the untitled file so that this window won't get reused
3. open terminal, run `export FOOBAR=42`
4. run `atom foodir` in terminal
5. open inspector in the last atom window, `process.env['FOOBAR']` is `"42"`

But:

- close window of step 2.
- completely quit atom
- restart atom from Dock, last window (of step 5) gets restored
- open inspector in the last atom window
   1. `process.env['FOOBAR']` is `undefined`
   2. `process.env['LANG']` (among other critical variables) is undefined
   3. `process.env['SHELL']` is set to `/bin/bash` when I'm using `/bin/zsh`

For those last two critical variables, methods equivalent to what's described down [this comment](https://github.com/atom/atom/issues/4126#issuecomment-132119777) should be used to set them from a launchd started instance (reproduced below for convenience):

```bash
LANG=$(osascript -e "user locale of (get system info)" | sed -e "s/@.*//" -e "s/\$/.UTF-8/")
SHELL="$(dscl . -read "/Users/$USER" UserShell | sed -e 's/UserShell: //')"
```

The code in the PR uses `-ilc`, among which is `-i`, meaning this spawns an *interactive* shell, which by definition the sync call isn't, which thus needlessly loads `.bashrc`, slowing down startup, and potentially doing things the user has put there that anticipates interactivity (such as ssh/gpg agent auth questions or fanciful colored output).

Additionally:

- close everything
- in one term, `export FOOBAR=42; atom foodir`
- in another term (possibly days later), `export FOOBAR=44; export FOOBAZ=46; atom foodir`
- the latter will merely make the opened window pop up and (obviously) not set vars into the window

In any case, although there are shared underlying issues, some of which have been fixed, *the "fixed use case" mentioned above isn't my use case*, in which I am in need of a uniform environment, whichever the way Atom has been started. IOW the use case of this issue is threefold:

- purposely disallow inheriting from a shell's environment (may be an application-wide setting)
- have a sane environment bootstrap available whichever way atom is started (notably `LANG` and `SHELL` vars)
- have an *option* to fetch environment from the user's login shell or not, and when it does, does it in a way that respects the ksh family `rc`/`profile` standard (`-i` vs `-l`)

Thanks for taking the time to listen, the work being poured into this is really appreciated.

**Comment by joefitzgerald** [comment]
I understand what you're saying. I think there are a few different potential issues to be opened based on this comment:

1. Should `-i` be used when spawning the shell? [Not sure]
1. Should `process.env` be serialized and reused when re-opening an existing project? [I can see how this would be useful]
1. Should `process.env` be updated if `atom` is invoked from the shell for the same directory multiple times? [I think yes]
1. [Bug] process.env['SHELL'] is set to /bin/bash when I'm using /bin/zsh

> [...] making each window a sandbox WRT process.env [...]

Of note, each `window` in atom _is actually a different process_ (i.e. it's a whole different node environment). The prior issue was that the environment was never being passed to new windows when launched from the terminal.

> [...] process.env['LANG'] (among other critical variables) is undefined [...]

This is probably because there is a bug in detecting your shell; causing the wrong shell to be invoked.

**Comment by lloeki** [comment]
> This is probably because there is a bug in detecting your shell; causing the wrong shell to be invoked.

No, because: 

- processes spawned from `launchd` don't get `LANG` (they're supposed to get it from OS X frameeworks)
- the shell (login or not) has no responsibility setting this on OS X† (not sure about `LaunchDaemons`/`LaunchAgents`). Instead, `Terminal.app` gets it from the OS framework and sets it *only* when the following preference is set in the Advanced tab of the profile. When unchecked, `LANG` is unset.

<img width="257" alt="screen shot 2016-03-23 at 17 28 58" src="https://cloud.githubusercontent.com/assets/75445/13992575/e9b7ca08-f11c-11e5-80b7-24a21d086662.png">

† neither in Linux (dunno about other *nix) for that matter, that's part of the init process (whether it's systemd or something else) and the shell (even when it's a login one) inherits it.
